### PR TITLE
[ISSUE #1515] SYNC_MASTER could be change into pipeline manner 

### LINKED ISSUE #1515: [improvement] SYNC_MASTER could be change into pipeline manner
** Performance improvement **

The performance of SYNC_MASTER is poor, only 1/10 of async.
According to benchmark
- Configure
    - CPU : Intel(R) Xeon(R) Gold 5118 CPU @ 2.30GHz, 48C
    - Memory : 512G
    - Disk : 2T PCIE SSD
- Result
    - ASYNC_MASTER : TPS = 108k; latency avg = 1.8ms; latency pct99=2.6ms
    - SYNC_MASTER : TPS = 10.6k; latency avg = 18.7ms; latency pct99=26ms

The cause of the poor performance of SYNC_MASTER is that the request processor blocking wait until the flushed offset larger than write offset, blocking the upcoming request.

This blocking wait manner can be changed into pipeline manner, when then send request is written to local commitLog, the processor can process the next request immediately; let the flushSlave/flushDisk thread responds to client the send request is replicated successfully or timeout.

### PR DESCRIPTION
**Pull request by shenhui0509**
this commit resolves #1515 change SYNC_MASTER flushSlave/flushDisk to pipeline manner

## What is the purpose of the change
performance improvement for SYNC_MASTER :
- reduce produce latency 
- improve produce throughput

## Brief changelog
1. add an interface *asyncProcessRequest* to remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRequestProcessor.java allow processors can process request in async manner, and by default wrap the result of ProcessRequest with CompletableFuture.
2. change remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java, let write response to client happen after the future is filled, rather than a blocking call
3. add and interface *asyncPutMessage* to MessageStore
4. change the store/src/main/java/org/apache/rocketmq/store/CommitLog.java, the *asyncPutMessage* method returns a CompletableFuture<PutRusult>, this future is filled by the flushSlave/flushDisk thread instead of the processor thread, thus the processor can process the next request just after append to local commitlog, let the flushSlave/flushDisk thread respond to client.
5. some dup code eliminate

### Functional
1. add unit test in store/src/test/java/org/apache/rocketmq/store/HATest.java
2. shutdown slave while sending message, the producer received the FLUSH_SLAVE_TIMEOUT response code.
image

### performance
benchmark config:
2 * broker：48C，512G mem，4 * 2T SSD; 1-master-1-slave
namesrv is hybrid deployed with broker
3 * client : 40C, 64G mem
message size : 1024B
write thread : 64

benchmark result 
*sync_pipeline* is the optimized version configured with SYNC_MASTER
*wait* is the origin version configured with SYNC_MASTER
*async* is the origin version configured with ASYNC_MASTER

TPS
image
image
Latency
image
image

The result shows that the optimized version's performance is close to ASYNC_MASTER.

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by shenhui0509** [comment]
ping @duhenglucky @vongosling

**Comment by duhenglucky** [comment]
@shenhui0509 good job, thanks for your contribution, we will review this PR ASAP.

**Comment by duhenglucky** [comment]
@shenhui0509 and it would be nice if you can pull this request to develop branch

**Comment by shenhui0509** [comment]
> @shenhui0509 and it would be nice if you can pull this request to develop branch

done

**Comment by shenhui0509** [comment]
Hi, @duhenglucky, is it necessary to write a RIP? If needed, I'd like to write one.

**Comment by xujianhai666** [comment]
> Hi, @duhenglucky, is it necessary to write a RIP? If needed, I'd like to write one.

@shenhui0509 I think this pr is so important, A RIP is necessary

**Comment by xujianhai666** [comment]
> Hi, @duhenglucky, is it necessary to write a RIP? If needed, I'd like to write one.

@shenhui0509 @duhenglucky I think this pr is so important, A RIP is necessary

**Comment by duhenglucky** [comment]
@xujianhai666 good suggestion, This PR is indeed a relatively big change, and on the critical path, it is best to write a RIP to describe the scope of the relevant changes, as well as the key design in this PR, if you have any other question, please feel free ask me directly .

File: remoting/pom.xml [hunk 1]
```diff
@@ -28,8 +28,8 @@
     <name>rocketmq-remoting ${project.version}</name>
 
     <properties>
-        <maven.compiler.source>1.6</maven.compiler.source>
-        <maven.compiler.target>1.6</maven.compiler.target>
+        <maven.compiler.source>1.8</maven.compiler.source>
```
Comment (vongosling) [inline remoting/pom.xml]: Do not modify compiler level. we need to keep to be compatible with JDK 1.6 in SDK

Comment (shenhui0509) [reply remoting/pom.xml]: I used some features in JDK 1.8. I'll try to resolve this compatible problem ASAP.

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java [hunk 2]
```diff
@@ -1041,26 +1327,39 @@ public long getJointime() {
...
         public boolean waitForFlush(long timeout) {
             try {
                 this.countDownLatch.await(timeout, TimeUnit.MILLISECONDS);
-                return this.flushOK;
-            } catch (InterruptedException e) {
+                return flushOk.get();
+            } catch (InterruptedException | ExecutionException e) {
                 log.error("Interrupted", e);
                 return false;
             }
```
[the remark this answers is not in the document]
Comment (shenhui0509) [reply store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: Ok, I eliminate that.

**Comment by shenhui0509** [comment]
> This PR does improve TPS and reduce latency when synchronized replication (Verified by benchmark). There are some issues：
> 
> 1. Transaction message does not seem to work after modification.
> 2. [DISCUSS] When slave down, master return FLUSH_SLAVE_TIMEOUT instead of SLAVE_NOT_AVALIABLE.

1. I missed msg properties when storage. 
2. I agree to return SLAVE_NOT_AVAILABLE.
Both 1 and 2 are fixed, and add some UT and IT for transaction message. Plz take a look at the new commit

**Comment by vongosling** [comment]
Very great optimization. I will pick some time to work with this pr in the next week.

**Pull request merged**